sips/sip-015/sip-015-network-upgrade.md

```diff
@@ -0,0 +1,1717 @@
+
+Even in PoX-1, delegation operators are capable of assigning multiple
+reward addresses for funds that they manage: each invocation of
+`delegate-stack-stx` can use a different PoX reward address. However,
+if operators wished to split the same _account's_ lockups across
+multiple addresses, this would not be possible.
+
+In PoX-2, however, delegation operators can use
+`delegate-stack-increase` to achieve this. This method allows the
+delegation operator to set a new PoX address to receive the
+"partially stacked" funds from the increase. Once called, the operator
+can use `stack-aggregation-commit` to commit to each PoX reward
+address separately.
+
+## Clarity
+
+If this SIP is ratified, then at the time of the
+Stacks 2.1 network upgrade, the Clarity smart
+contract language will be expanded with new features and some new
+behaviors for existing native methods. In order to support these
+changes, this SIP proposes to introduce *Clarity versioning for smart
+contracts*. In this scheme, each smart contract will be associated
+with a particular Clarity version. The execution environment will
+track the current version for a given execution (in the
+implementation, this will be via the `ContractContext`), and use that
+to select which features are available, and which native method
+implementations will be used. Clarity 2 contracts can invoke Clarity 1
+contracts (and vice-versa), but particular care will need to be taken
+if a new native keyword is used in the Clarity 1 contract's API.
+
+For example:
+
+```
+Contract A (Clarity 1 Contract):
+  (define-public (stx-account) ...)
+  
+Contract B (Clarity 2 Contract):
+  (contract-call contract-A stx-account)
+```
+
+In such cases, the new keyword (in the above example, the keyword is
+`stx-account`) *cannot* be used by the Clarity 2 contract, and
+therefore invoking the public method of Contract A from Contract B is
+not allowed. To address these cases, contract authors must launch a
+Clarity 1 contract that interposes on Contract A, providing a Clarity
+2 compatible interface.
+
+Existing, pre-2.1 contracts will all be Clarity 1. New contracts will
+default to Clarity 2, and a new Stacks transaction wire format for publishing
+smart contracts will allow contract publishers to choose the version that
+their contract should use.
+
+Note that the act of adding, changing, or removing a native Clarity function or native
+Clarity global variable necessitates the creation of a new version of the
+Clarity language, and must be treated as a breaking change.
+This is because adding, changing, or removing either of these
+things alters the rules for block validation, which makes these 
+consensus-level changes.  This SIP proposes introducing a new version of Clarity
+(Clarity 2) _while also_ preserving the current version of Clarity (Clarity 1).  This
+shall not be construed as setting a precedent -- a future SIP may remove the
+ability to publish new smart contracts with older versions of Clarity.
```

Comment:
The SIP should discuss how the version scheme affects _traits_. 

A trait defined in one version of Clarity may not be possible to implement in a later version. For example, say a Clarity-1 contract defines a trait with a function `slice`: 

```clarity
(define-trait my-trait  
  ((slice ((list 9 int) (offset int) (length int)) 
     (list 9 int))))
```

As `slice` is a native function in Clarity-2, the trait can only be implemented in Clarity-1 contracts. Worse, the trait may not even be possible to implement and deploy in future releases of the Stacks blockchain:

> a future SIP may remove the ability to publish new smart contracts with older versions of Clarity.

Reply:
This is explained in the intro to the `Clarity` subsection of the `Specification` section.  In order to call Clarity 1 from Clarity 2, Clarity 2 keywords cannot be present in a Clarity 1's contract API.

Reply:
It's a big deal that _traits aren't forward compatible_, and should be explicitly stated in the SIP, not brushed under the rug. 

Here is another concrete example of the problems with the version scheme: SIP-009 and SIP-010 each define a trait with a `transfer` function. If a future version of Clarity introduces a native function named `transfer`, these standard traits for tokens cannot be implemented in that or later versions of Clarity, if at all.

Reply:
Do you know of any traits that would be affected?

We didn't add any keywords that collided with known traits' APIs, but that could have changed since they were written.

Reply:
All traits are affected by not being guaranteed to be implementable in a later Clarity version than they were defined.

Reply:
Let me rephrase -- as of right now, do you know of any trait definitions whose function names collide with any of the proposed Clarity 2 keywords?  If the answer is "no," then this isn't really a concern.

Reply:
Burying the head in the sand doesn't make the problem with traits go away. It doesn't matter at all what I know about the different trait definitions out there. Traits not being forward compatible is a no-go.

Reply:
> Burying the head in the sand doesn't make the problem with traits go away.

No one is pretending that the possibility of a name collision doesn't exist between now and the proposed activation height.  The choice is between the following options:

* Rewrite substantial portions of the VM's type system to permit using reserved words as function names (which has its own problems -- namely, masking built-ins makes it harder to reason about what a code snippet actually does).  This SIP is not proposing this, and I doubt any of the SIP authors have an interest in undertaking this.

* Implement new reserved words that are unlikely to collide with existing public functions.

To assess the risk of a name collision creating problems for users who need Clarity 1/2 interoperability, I went ahead and loaded up _ever single smart contract_ on the chain and searched for _any_ collision with these new keywords.  Turns out, a bunch of contracts had simply copy/pasted my original Clarity 1 `buff-to-uint-le` function, and that's it.  My original `buff-to-uint-le` function has the _exact same behavior_ as the Clarity 2 `buff-to-uint-le` function, so new Clarity 2 code will _not_ need to call into a Clarity 1 implementation of `buff-to-uint-le`.

So while name collisions are _possible_, the fall-out appears _inconsequential_.

Reply:
You can see for yourself:

```bash
$ sqlite3 ./mainnet/chainstate/vm/clarity/marf.sqlite 'select value from metadata_table where value like "%define-public%"' >> /tmp/functions.txt
$ sqlite3 ./mainnet/chainstate/vm/clarity/marf.sqlite 'select value from metadata_table where value like "%define-read-only%"' >> /tmp/functions.txt
$ cat /tmp/words         
stx-account                                       
principal-destruct                                
principal-construct                               
get-burn-block-info?                              
slice                                               
string-to-int                                     
string-to-uint                                    
int-to-ascii                                      
int-to-utf8                                       
buff-to-int-le                                    
buff-to-uint-le                                   
buff-to-int-be                                    
buff-to-uint-be                                   
stx-transfer-memo?                                
is-standard                                       
to-consensus-buff                                 
from-consensus-buff                               
replace-at
$ for word in $(cat /tmp/words); do grep "$word" /tmp/functions.txt; done
```

It's only about 1400 lines to trawl through, but you can break it up by keyword to verify that the only publicly-callable function affected is `buff-to-uint-le`.

Reply:
> Here is another concrete example of the problems with reserving keywords and the version scheme: [SIP-009](https://github.com/stacksgov/sips/blob/main/sips/sip-009/sip-009-nft-standard.md) and [SIP-010](https://github.com/stacksgov/sips/blob/main/sips/sip-010/sip-010-fungible-token-standard.md) each define a trait with a `transfer` function. If a future version of Clarity introduces a native function named `transfer`, a SIP-009 or SIP-010 compliant smart contract cannot be implemented in that or later versions of Clarity.

I don't understand this argument. A future version of Clarity is going to be decided upon by the community. Why would the community introduce something that breaks a standard codified in a SIP? This is not a real problem.

Reply:
Agreed.  If there was a SIP that proposed making `transfer` a native keyword, and in doing so, broke SIP-009 and SIP-010, and somehow the SIP advanced to Activation-in-Progress status, the community would vote it down.